tests/test_python3_minus_m_pip_in_gh_actions.py

```diff
@@ -0,0 +1,82 @@
+"""
+Ensure that "python3 -m pip install " is used consistantly in GitHub Actions.
+
+Find all .yml files in the .github/workflows directory and find all lines that contain
+" pip install " and ensure that they also contain "python3 -m pip install ".
+
+If there is any line that has the first but not the second, then print the file name,
+line number, and the line.  Raise an error if any non-compliant files are found.
+"""
+
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Iterable
+from unittest import mock
+
+import pytest
+
+
+def process_file(file_path: Path) -> Iterable[tuple[int, str]]:
+    """Process a file and yield any lines that contain " pip install " but not
+    "python3 -m pip install ".
+    """
+    with file_path.open() as in_file:
+        for line_number, line in enumerate(in_file, start=1):
+            if " pip install " in line and "python3 -m pip install " not in line:
+                yield line_number, line.strip()
+
+
+def test_pip_install_without_python3_minus_m_prefix() -> None:
+    workflows_dir = Path(".github/workflows")
+    if not workflows_dir.is_dir():
+        msg = f"Directory {workflows_dir} does not exist."
+        raise FileNotFoundError(msg)
+
+    errors = []
+    for yml_file in workflows_dir.glob("*.yml"):
+        errors.extend(
+            f"{yml_file}:{line_number}: {line}"
+            for line_number, line in process_file(yml_file)
+        )
```

Comment:
Let's simplify this - just because you can doesn't mean you should.  The indenting here makes this particularly bad as a construct!



```suggestion
        for line_number, line in process_file(yml_file):
            errors.append(f"{yml_file}:{line_number}: {line}")
```

Reply:
The proposed solution reduces function call overhead from once per failing line to once per failing file.

Runtime performance is an important consideration for code that we may choose to add as a pre-commit hook.

% `ruff rule FURB113 ` # [repeated-append](https://docs.astral.sh/ruff/rules/repeated-append/#repeated-append-furb113)
[ ... ]
[Why is this bad?](https://docs.astral.sh/ruff/rules/repeated-append/#why-is-this-bad)

Consecutive calls to append can be less efficient than batching them into a single extend. Each append resizes the list individually, whereas an extend can resize the list once for all elements.
[ ... ]